feat: audit tracing improvement

kzamlynska opened the pull request:
closes #344

ludwiktrammer commented on packages/ragbits-core/src/ragbits/core/audit/base.py:
```diff
@@ -88,7 +88,10 @@ def trace(self, name: str, **inputs: Any) -> Iterator[SimpleNamespace]:  # noqa:
 
 
-def format_attributes(data: dict, prefix: str | None = None) -> dict:
+# TODO: check do we need this method.
```
What does this mean? Check how?

kzamlynska replied:
I've checked it. I needed to run the otel tracer and see how the new formater looks like.
I removed this method.

kzamlynska replied:
After discussion I restored the format_attributes the way it was for otel trace handler .

ludwiktrammer commented on packages/ragbits-core/src/ragbits/core/audit/base.py:
```diff
@@ -119,3 +122,112 @@ def format_attributes(data: dict, prefix: str | None = None) -> dict:
+
+    def shorten_list(lst: list | tuple) -> list:
+        """
+        Shortens a list if it's longer than 3 elements. Shortens list elements if it's long string.
```
```suggestion
        Shortens a list if it's longer than opt_list_lenght elements. Shortens list elements if it's long string.
```

kzamlynska replied:
changed

ludwiktrammer commented on packages/ragbits-core/src/ragbits/core/audit/base.py:
```diff
@@ -119,3 +122,112 @@ def format_attributes(data: dict, prefix: str | None = None) -> dict:
     return flattened
+
+
+def format_attributes(data: dict, prefix: str | None = None) -> dict:
```
This function with nested sub-functions that use local variables from the parent function pretty much looks like simulating a class with methods and instance properties. Maybe simply use a class if a class-like functionality is needed?

kzamlynska replied:
I've changed the function to the class

kzamlynska commented:
I've added @traceable decorator to the most of the methods in the ragbits.core 
I've added @traceable decorator to some of the methods in other ragbits packages - (I've chosen the one with more complicated logic)
I've omitted ragbits.core.audit i ragbits.core.utils -> please let me know if we need tracing in the second one.
I've ommitted ragbits.cli

micpst reviewed (commented):
I think the main problems of the #344 issue are not solved in this PR. I'm talking about:

> 
> 1. Review the main pathways in our library, identify areas that are not currently traced, and add tracing to those sections.
> 
> 2. Sometimes we send traces that should not be included in the tracing event, such as vectors. We need to eliminate these from the traces.
> 

The main goal was to replace `traceable` decorators with `track` contexts with filtered arguments that are actually meaningful to the user, such as removing vector values, etc.

I don't see it in this PR, and I also feel that too much tracking has been added, such as tracking in private methods, which are irrelevant to the user.

kzamlynska commented:
If you want to see the newest prompt/response parsing please use examples/core/llm.py
